Documentation/gen-api.md

```diff
@@ -0,0 +1,26 @@
+# GenApi Usage in WPF on .NET Core
+In WPF on .NET Core, C# reference assemblies are created via the use of [GenAPI](https://github.com/dotnet/arcade/tree/master/src/Microsoft.DotNet.GenAPI) and a separate reference assembly project located in the `ref` directory under a particular assemblies source directory.
+
+WPF assemblies make extensive use of the [InternalsVisibleToAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute?view=netcore-3.0) which precludes the use of [ProduceReferenceAssembly](https://docs.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-properties?view=vs-2019) or [ProduceOnlyReferenceAssembly](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/refonly-compiler-option).  This is because these compiler options will include internal types and members in the reference assembly.  In WPF, this creates dangling references to assemblies that do not exist in the `WindowsDesktop` reference pack.
+
+Using GenAPI allows us to strip out internals, removing the dangling references from our reference assemblies.
+
+## Using GenAPI in WPF
+GenAPI is run only on-demand.  In the event that a change to a runtime assembly creates new public surface area, a developer will see an [ApiCompat](api-compat.md) error between the reference assembly and the runtime assembly.  In order to address this, the developer must run GenAPI to generate new reference assembly code.
+### Running GenAPI
+GenAPI can be run by setting the following MSBuild property while building.
+```
+/p:GenerateReferenceAssemblySource=true
+```
+When a build is run with that property enabled, GenAPI will read the runtime assembly and generate a new `{AssemblyName}.cs` file under the ref directory in the assembly's source tree.
+
+This new file will contain the newly created surface area and will need to be checked in along with the runtime assembly change.  The next build without `GenerateReferenceAssemblySource` enabled will no longer display an ApiCompat error as the surface area will now match the baseline.
+### Issues with GenAPI
```

Comment:
is all of this extra work needed because we use the `InternalsVisibleToAttribute`?

Reply:
It's a combination of a few things, `InternalsVisibleTo` certainly complicates things between assemblies, but also XAML compilation will reflect and try to find internal API in order to take certain shortcuts and optimizations.  (Look at [IAddChildInternal](https://github.com/dotnet/wpf/search?q=IAddChildInternal&unscoped_q=IAddChildInternal) for example).

There are also issues where we have public API that has internal abstract members, which confuses GenAPI.  GenAPI itself just sometimes adds strange things, like private dummy int fields and properties.  No idea why, there are just bugs.  I intend to make small reproduction applications when I have time and file bugs against GenAPI, but that's future stuff.

Reply:
Are those bugs reported back to the Roslyn team so they can be fixed?

Reply:
GenAPI doesn't use Roslyn it uses [Microsoft.CCI](https://github.com/Microsoft/cci)).  Originally, we were using Roslyn to generate our reference assemblies, but the use of InternalsVisibleTo was creating dangling references.  We filed https://github.com/dotnet/roslyn/issues/36409 to address these issues.  In the meantime, this was the fix.

I haven't created issues against GenAPI yet, I haven't had the time to create minimal reproductions of the various bugs.

Reply:
If i'm adding a new function to a class, what's the difference between adding the API to the reference assembly manually and using GenAPI? I might not fully understand the problem we are trying to solve here, but it seems like it would be much simpler to just manually modify the reference assembly. Is there some part of our workflow that makes things more complicated?

Reply:
I missed this.  The C# reference assemblies didn't exist in a manual fashion before this (back in the old internal repo maybe), so we could not just add things.  So GenAPI was used as an initial seed for the most part.  For some changes, it's probably easy enough just to pop the change into the reference assembly.  Other changes it might be more difficult and require basically mirrored code changes.  Also if you make a small change sans GenAPI then someone makes a very large change and uses it, it'll likely generate a noisier diff since your code location might not be the same.

It's probably the simplest and most thorough to GenAPI and use that output to manually add to the reference assembly.  It's pretty easy (mostly) to diff the generated output against the current surface and just pick and choose your new stuff.

Long term we'd rather just have Roslyn allow us to say, "All publics, but also these specific internals" in a nice configurable fashion and call it a day.  But we need something in the short term.